core/evm.go

```diff
@@ -27,17 +27,71 @@
 package core
 
 import (
+	"bytes"
 	"math/big"
 
 	"github.com/ava-labs/coreth/consensus"
 	"github.com/ava-labs/coreth/consensus/misc/eip4844"
+	"github.com/ava-labs/coreth/core/extstate"
+	"github.com/ava-labs/coreth/core/state"
 	"github.com/ava-labs/coreth/core/types"
-	"github.com/ava-labs/coreth/core/vm"
-	"github.com/ava-labs/coreth/predicate"
-	"github.com/ethereum/go-ethereum/common"
+	"github.com/ava-labs/coreth/params"
+	"github.com/ava-labs/coreth/params/extras"
+	customheader "github.com/ava-labs/coreth/plugin/evm/header"
+	"github.com/ava-labs/libevm/common"
+	ethtypes "github.com/ava-labs/libevm/core/types"
+	"github.com/ava-labs/libevm/core/vm"
 	"github.com/holiman/uint256"
 )
 
+func init() {
+	vm.RegisterHooks(hooks{})
+}
+
+type hooks struct{}
+
+// OverrideNewEVMArgs is a hook that is called back when a new EVM is created.
+// It allows for the modification of the EVM arguments before the EVM is created.
+// Specifically, we set Random to be the same as Difficulty since Shanghai.
+// This allows using the same jump table as upstream.
+// Then we set Difficulty to 0 as it is post Merge in upstream.
+// Additionally we wrap the StateDB with the appropriate StateDB wrapper,
+// which is used in coreth to process historical pre-AP1 blocks with the
+// GetCommittedState method as it was historically.
+func (hooks) OverrideNewEVMArgs(args *vm.NewEVMArgs) *vm.NewEVMArgs {
+	rules := args.ChainConfig.Rules(args.BlockContext.BlockNumber, params.IsMergeTODO, args.BlockContext.Time)
+	args.StateDB = wrapStateDB(rules, args.StateDB)
+
+	if rules.IsShanghai {
+		args.BlockContext.Random = new(common.Hash)
+		args.BlockContext.Random.SetBytes(args.BlockContext.Difficulty.Bytes())
+		args.BlockContext.Difficulty = new(big.Int)
+	}
+
+	return args
+}
+
+func (hooks) OverrideEVMResetArgs(rules params.Rules, args *vm.EVMResetArgs) *vm.EVMResetArgs {
```

Comment:
nit since the `args` are passed by pointer and modified in place, should we NOT return the argument? It does look a bit strange from the signature to return the evmresetargs although the original has been modified as well. 🤔